vortex-expr/src/pack.rs

```diff
@@ -0,0 +1,233 @@
+use std::any::Any;
+use std::fmt::Display;
+use std::sync::Arc;
+
+use itertools::Itertools as _;
+use vortex_array::aliases::hash_set::HashSet;
+use vortex_array::array::StructArray;
+use vortex_array::validity::Validity;
+use vortex_array::{ArrayData, IntoArrayData};
+use vortex_dtype::field::Field;
+use vortex_dtype::FieldNames;
+use vortex_error::{vortex_bail, VortexResult};
+
+use crate::{ExprRef, VortexExpr};
+
+#[derive(Debug, Clone)]
+pub struct Pack {
+    names: FieldNames,
+    values: Vec<ExprRef>,
+    validity: Option<ExprRef>,
+}
+
+impl Pack {
+    pub fn try_new_expr(
+        names: FieldNames,
+        values: Vec<ExprRef>,
+        validity: Option<ExprRef>,
+    ) -> VortexResult<Arc<Self>> {
+        if names.len() != values.len() {
+            vortex_bail!("length mismatch {} {}", names.len(), values.len());
+        }
+        if names.len() < 1 {
+            vortex_bail!("must provide at least one field");
+        }
+        Ok(Arc::new(Pack {
+            names,
+            values,
+            validity,
+        }))
+    }
+}
+
+impl PartialEq<dyn Any> for Pack {
+    fn eq(&self, other: &dyn Any) -> bool {
+        other.downcast_ref::<Pack>().is_some_and(|other_pack| {
+            self.names == other_pack.names
+                && self
+                    .values
+                    .iter()
+                    .zip(other_pack.values.iter())
+                    .all(|(x, y)| x.eq(y))
+        })
+    }
+}
+
+impl Display for Pack {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Pack(")?;
```

Comment:
Why not use struct formatters (debug_struct)? I think you can use it for display too